curp/src/sync_manager.rs

```diff
@@ -144,137 +137,222 @@ impl<C: Command + 'static> SyncManager<C> {
+        // TODO: gracefully stop the background tasks
+        let _background_ping_handle = tokio::spawn(Self::bg_append_entries(
+            self.connects(),
+            Arc::clone(&self.state),
+            Arc::clone(&commit_notify),
+            Arc::clone(&ae_notify),
+        ));
+        let _background_apply_handle = tokio::spawn(Self::background_apply(
+            Arc::clone(&self.state),
+            cmd_executor,
+            comp_chan,
+            Arc::clone(&commit_notify),
+        ));
 
         loop {
             let (cmds, term) = match self.fetch_sync_msgs().await {
                 Ok((cmds, term)) => (cmds, term),
                 Err(_) => return,
             };
 
-            let index = self.log.map_lock(|mut log| {
-                let len = log.len();
-                log.push(LogEntry::new(term, &cmds, EntryStatus::Unsynced));
-                len
+            self.state.map_write(|mut state| {
+                state.log.push(LogEntry::new(term, &cmds));
+
+                debug!(
+                    "received new log, index {:?}",
+                    state.log.len().checked_sub(1)
+                );
             });
-            let cmds_arc: Arc<[_]> = cmds.into();
+        }
+    }
 
-            let comp_fn = || {
-                for c in cmds_arc.iter() {
-                    if self
-                        .comp_chan
-                        .send(
-                            c.keys(),
-                            SyncCompleteMessage::new(index.numeric_cast(), Arc::clone(c)),
-                        )
-                        .is_err()
-                    {
-                        error!("The comp_chan is closed on the remote side");
-                        return false;
-                    }
-                }
-                true
-            };
+    /// Heartbeat Interval
+    const APPEND_ENTRIES_INTERVAL: Duration = Duration::from_millis(150);
+    /// Heartbeat Timeout
+    const APPEND_ENTRIES_TIMEOUT: Duration = Duration::from_millis(50);
 
-            if !Self::broadcast_and_wait(
-                self.connects(),
-                iter::repeat_with(|| Arc::clone(&cmds_arc)),
-                |(connect, cmds_cloned)| async move {
-                    connect
-                        .sync(
-                            SyncRequest::new(term, index.numeric_cast(), cmds_cloned.as_ref())?,
-                            SYNC_TIMEOUT,
+    /// Background ping, only work for the leader
+    async fn bg_append_entries(
+        connects: Vec<Arc<Connect>>,
+        state: Arc<RwLock<State<C>>>,
+        commit_notify: Arc<Notify>,
+        ae_notify: Arc<Notify>,
+    ) {
+        #[allow(clippy::integer_arithmetic)] // tokio internal triggered
+        loop {
+            // wait for either heartbeat timeout
+            tokio::select! {
+                _ = tokio::time::sleep(Self::APPEND_ENTRIES_INTERVAL) => {},
+                _ = ae_notify.notified() => {},
```

Comment:
We should not send heartbeat once an entry getting committed. We should reset the heartbeat timeout once we get a notification.